diff --git a/taxonomy_single_tag.module b/taxonomy_single_tag.module
index bae0dea..6df1fbe 100644
--- a/taxonomy_single_tag.module
+++ b/taxonomy_single_tag.module
@@ -68,7 +68,8 @@ function taxonomy_single_tag_field_widget_form(&$form, &$form_state, $field, $in
     '#element_validate' => array('taxonomy_single_tag_autocomplete_validate'),
   );
 
-  return $element;
+  // Return in tid key, so we could replace values in validation stage
+  return array('tid' => $element);
 }
 
 /**
@@ -80,7 +81,10 @@ function taxonomy_single_tag_field_widget_form(&$form, &$form_state, $field, $in
 function taxonomy_single_tag_autocomplete_validate($element, &$form_state) {
   // Autocomplete widgets do not send their tids in the form, so we must detect
   // them here and process them independently.
-  $value = '';
+  $value = array();
+  // Cause we are replacing whole delta item values,
+  // we need to set tid as empty string here. @see taxonomy_field_is_empty()
+  $value['tid'] = '';
   if ($tag = $element['#value']) {
     // Collect candidate vocabularies.
     $field = field_widget_field($element, $form_state);
@@ -96,6 +100,12 @@ function taxonomy_single_tag_autocomplete_validate($element, &$form_state) {
     // otherwise, create a new 'autocreate' term for insert/update.
     if ($possibilities = taxonomy_term_load_multiple(array(), array('name' => trim($tag), 'vid' => array_keys($vocabularies)))) {
       $term = array_pop($possibilities);
+
+      // Copy weight from input to element, so they will be sorted. If we don't do it they
+      // won't change delta if weight changed.
+      if (!empty($form_state['input'][$element['#field_name']][$element['#language']][$element['#delta']]['_weight'])) {
+        $term->_weight = $form_state['input'][$element['#field_name']][$element['#language']][$element['#delta']]['_weight'];
+      }
     }
     else {
       $vocabulary = reset($vocabularies);
@@ -108,6 +118,9 @@ function taxonomy_single_tag_autocomplete_validate($element, &$form_state) {
     }
     $value = (array) $term;
   }
+  // Replace whole delta item values with normal term objects.
+  $element['#parents'] = array($element['#field_name'], $element['#language'], $element['#delta']);
+
   form_set_value($element, $value, $form_state);
 }
 
@@ -159,4 +172,4 @@ function taxonomy_single_tag_autocomplete($field_name, $tag_typed = '') {
   }
 
   drupal_json_output($term_matches);
-}
\ No newline at end of file
+}
